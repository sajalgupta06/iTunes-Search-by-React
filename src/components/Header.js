/* global $ */
// @flow
/* eslint-disable jsx-a11y/anchor-is-valid, jsx-a11y/click-events-have-key-events */
import React from 'react';
import emitter from '../emitter';
import type { HeaderState, SearchOption } from '../type';
import '../style/Header.css';

const options: Array<SearchOption> = [
  'All',
  'Audiobook',
  'eBook',
  'Movie',
  'Music',
  'Music Video',
  'Podcast',
  'TV Show',
  'Software'
];

class Header extends React.PureComponent<{}, HeaderState> {
  emitSearch: () => void;
  _onKeyUp: (e: Object) => void;
  _onClick: (e: Object) => void;
  _update: (e: Object) => Function;
  ticking: boolean;
  rAf: any;

  state: HeaderState = {
    media: 'All',
    query: ''
  };

  constructor(props: Object) {
    super(props);
    this.ticking = false;
    this.rAf = null;
    this.emitSearch = () => emitter.emit('search', this.state);
    this._onClick = e => this.setState(
      { media: e.target.textContent },
      () => (this.state.query.length ? this.emitSearch() : null)
    );
    this._update = ({ keyCode, target: { value: query } }) => _ => {
      this.setState({ query }, () => keyCode === 13 && this.emitSearch());
      this.ticking = false;
    };
    this._onKeyUp = e => {
      if (!this.ticking) {
        this.rAf = window.requestAnimationFrame(this._update(e));
        this.ticking = true;
      }
    };
  }

  renderSearchOption = () => options.map(opt =>
    <li
      key={opt}
      className={this.state.media === opt ? 'select' : ''}
    >
      <a
        tabIndex="0"
        role="button"
        onClick={this._onClick}
      >
        {opt}
      </a>
    </li>);

  componentWillUnmount() {
    window.cancelAnimationFrame(this.rAf);
  }

  render() {
    return (
      <div className="navbar-fixed">
        <nav>
          <div className="nav-wrapper cyan">
            <div className="header-search-wrapper">
              <i className="material-icons black-text">search</i>
              <input
                type="text"
                placeholder="Search"
                onKeyUp={this._onKeyUp}
                className="search-input white black-text"
              />
            </div>
            <div className="dropdown-wrapper">
              <div
                className="dropdown-button"
                role="button"
                data-activates="search-options"
              >
                {this.state.media}
              </div>
              <ul id="search-options" className="dropdown-content">
                {this.renderSearchOption()}
              </ul>
            </div>
          </div>
        </nav>
      </div>
    );
  }
}

export default Header;
