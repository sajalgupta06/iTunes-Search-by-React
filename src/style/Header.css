.dropdown-wrapper {
  text-align: center;
  position: relative;
  width: 100px;
  left: 60px;
  display: block;
  padding: 0 10px;
}

.dropdown-wrapper:hover {
  background: #009688;
}

.nav-wrapper {
  display: flex;
  justify-content: center;
}

.header-search-wrapper {
  position: relative;
  width: 30vw;
}

.header-search-wrapper i {
  position: absolute;
  left: 12px;
}

.header-search-wrapper input.search-input {
  background: rgba(255, 255, 255, .3);
  border-radius: 3px;
  padding-left: 50px;
}

.btn-wrapper {
  margin: auto 10px;
}

ul li.select {
  font-weight: 600;
  background: #e0e0e0;
}
