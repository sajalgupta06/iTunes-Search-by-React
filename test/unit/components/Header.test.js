import React from 'react';
import { shallow } from 'enzyme';
import Header from '../../../src/components/Header';

describe('<Header />', () => {
  let component;
  let instance;
  let mockEvent;
  beforeEach(() => {
    mockEvent = {
      keyCode: 13,
      target: {
        value: 'test'
      }
    };
    component = shallow(<Header />);
    instance = component.instance();
  });
  test('Header component should renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
  test('_onKeyUp should start requestAnimationFrame', () => {
    instance._onKeyUp(mockEvent);
    expect(instance.ticking).toBeTruthy();
    expect(instance.rAf).toBeDefined();
  });
  test('_update should set new state', () => {
    instance._update(mockEvent)();
    expect(component.state().query).toBe('test');
    expect(instance.ticking).toBeFalsy();
  });
});
