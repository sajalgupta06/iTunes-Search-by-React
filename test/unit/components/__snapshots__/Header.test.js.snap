// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Header component should renders correctly 1`] = `
<div
  className="navbar-fixed"
>
  <nav>
    <div
      className="nav-wrapper cyan"
    >
      <div
        className="header-search-wrapper"
      >
        <i
          className="material-icons black-text"
        >
          search
        </i>
        <input
          className="search-input white black-text"
          onKeyUp={[Function]}
          placeholder="Search"
          type="text"
        />
      </div>
      <div
        className="dropdown-wrapper"
      >
        <div
          className="dropdown-button"
          data-activates="search-options"
          role="button"
        >
          All
        </div>
        <ul
          className="dropdown-content"
          id="search-options"
        >
          <li
            className="select"
            key="All"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              All
            </a>
          </li>
          <li
            className=""
            key="Audiobook"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              Audiobook
            </a>
          </li>
          <li
            className=""
            key="eBook"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              eBook
            </a>
          </li>
          <li
            className=""
            key="Movie"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              Movie
            </a>
          </li>
          <li
            className=""
            key="Music"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              Music
            </a>
          </li>
          <li
            className=""
            key="Music Video"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              Music Video
            </a>
          </li>
          <li
            className=""
            key="Podcast"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              Podcast
            </a>
          </li>
          <li
            className=""
            key="TV Show"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              TV Show
            </a>
          </li>
          <li
            className=""
            key="Software"
          >
            <a
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              Software
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
